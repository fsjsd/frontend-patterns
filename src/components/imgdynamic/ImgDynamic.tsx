/* eslint-disable jsx-a11y/alt-text */
import React, { useRef } from 'react'
import styled, { keyframes } from 'styled-components';
import { useEffectAsync } from '../../utils/hooks/useEffectAsync';
import { useIsOnScreen } from '../../utils/hooks/useIsOnScreen';
import { imagePromise } from '../../utils/imagePromise';

interface ImgDynamic extends React.DetailedHTMLProps<React.ImgHTMLAttributes<HTMLImageElement>, HTMLImageElement> {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  loadingComp: React.ComponentElement<any, any>;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  errorComp: React.ComponentElement<any, any>;
}

const fadeInAnimation = keyframes`
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
`;

const Wrap = styled.div`
  &> img {
    animation: ${fadeInAnimation} ease 2s;
    animation-iteration-count: 1;
    animation-fill-mode: forwards;
    z-index: -1;
  }
`;

/**
 * Dynamic image component. Does not load image until it's within viewport. Additionally,
 * allows custom loading to be provided while image is loading and custom error component
 * to be displayed if image load fails.
 * @param param0 props
 * @returns 
 */
const ImgDynamic = ({ loadingComp, errorComp, ...props }: ImgDynamic) => {

  const [loadState, setLoadState] = React.useState<boolean | undefined>(undefined);
  const wrapperRef = useRef(null);
  const isOnScreen = useIsOnScreen(wrapperRef);

  useEffectAsync(async () => {
    if (!props.src || !isOnScreen)
      return;

    try {
      await imagePromise(props.src);
      setLoadState(true);
    } catch (e) {
      setLoadState(false);
    }
  }, [props.src, isOnScreen])

  // Enhancement for later: Always render image with visibility/display hidden,
  // then face in once on screen. Could potentially use a ref and mount the imagePromise
  // response element to the dom.

  return (<Wrap ref={wrapperRef}>
    {loadState === undefined && loadingComp}
    {loadState === false && errorComp}
    {loadState === true && <img {...props} />}
  </Wrap>
  )
}

export default ImgDynamic