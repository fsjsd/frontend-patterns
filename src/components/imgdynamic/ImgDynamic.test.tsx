/* eslint-disable jsx-a11y/label-has-associated-control */
import React from 'react'
import { render, waitFor } from '@testing-library/react'
import ImgDynamic from './ImgDynamic';

const loading = <div>loading</div>
const error = <div>error</div>
const imageSize = 200;

const testImg = "data:image/gif;base64,R0lGODlhEAAQAMQAAORHHOVSKudfOulrSOp3WOyDZu6QdvCchPGolfO0o/XBs/fNwfjZ0frl3/zy7////wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACH5BAkAABAALAAAAAAQABAAAAVVICSOZGlCQAosJ6mu7fiyZeKqNKToQGDsM8hBADgUXoGAiqhSvp5QAnQKGIgUhwFUYLCVDFCrKUE1lBavAViFIDlTImbKC5Gm2hB0SlBCBMQiB0UjIQA7";

const mockNativeLazyLoadSupportedGetter = jest.fn();
jest.mock('./utils', () => ({
  get NATIVE_LAZYLOAD_SUPPORTED() {
    return mockNativeLazyLoadSupportedGetter();
  },
}));

function setupImageMock(raiseOnLoadEvent: boolean, raiseOnErrorEvent: boolean) {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (globalThis as any).Image = function () {
    const img = {
      onload,
      onerror,
    } as HTMLImageElement;

    if (raiseOnLoadEvent) {
      setTimeout(() => {
        img.onload && img.onload({} as Event);
      }, 10);
    }

    if (raiseOnErrorEvent) {
      setTimeout(() => {
        img.onerror && img.onerror({} as Event);
      }, 10);
    }
    return img
  }
}

describe("ImgDynamic", () => {

  let mockIntersectionObserver;
  beforeEach(() => {
    jest.resetAllMocks();
    mockIntersectionObserver = jest.fn();
    mockIntersectionObserver.mockReturnValue({
      observe: () => null,
      unobserve: () => null,
      disconnect: () => null
    });
    window.IntersectionObserver = mockIntersectionObserver;
  })

  it("renders loading while image downloading", async () => {
    const { getByText } = render(<ImgDynamic
      id={`imgId`}
      loadingComp={loading}
      errorComp={error}
      width={imageSize}
      height={imageSize}
      src={`loading ...`}
    />)
    await waitFor(() => expect(getByText("loading")).toBeInTheDocument());
  });

  it("renders image correctly on legacy browsers", async () => {
    setupImageMock(true, false);

    const { container, queryByText } = render(<ImgDynamic
      id={`imgId`}
      loadingComp={loading}
      errorComp={error}
      width={imageSize}
      height={imageSize}
      src={testImg}
    />)

    expect(mockIntersectionObserver).toHaveBeenCalledTimes(1);
    const calls = mockIntersectionObserver.mock.calls;
    const [onChange] = calls[calls.length - 1];
    onChange([{ isIntersecting: true, intersectionRatio: 0.5 }]);

    await waitFor(() => expect(queryByText("error")).not.toBeInTheDocument());

    await waitFor(() =>
      expect(container.querySelector('#imgId')).toHaveStyle("visibility: visible")
    );

    expect(container).toMatchSnapshot();
  })

  it.skip("renders image correctly", async () => {
    mockNativeLazyLoadSupportedGetter.mockReturnValue(true);
    setupImageMock(true, false);

    const { container, queryByText } = render(<ImgDynamic
      id={`imgId`}
      loadingComp={loading}
      errorComp={error}
      width={imageSize}
      height={imageSize}
      src={testImg}
    />)

    expect(mockIntersectionObserver).not.toHaveBeenCalled();

    await waitFor(() => expect(queryByText("error")).not.toBeInTheDocument());
    await waitFor(() => expect(queryByText("loading")).not.toBeInTheDocument(), { timeout: 100 });
    await waitFor(() =>
      expect(container.querySelector('#imgId')).toHaveStyle("visibility: visible")
    );

    expect(container).toMatchSnapshot();
  })

  it("does not render image off screen", async () => {
    setupImageMock(true, false);
    mockNativeLazyLoadSupportedGetter.mockReturnValue(false);

    const { container, queryByText } = render(<ImgDynamic
      id={`imgId`}
      loadingComp={loading}
      errorComp={error}
      width={imageSize}
      height={imageSize}
      src={testImg}
    />)

    expect(mockIntersectionObserver).toHaveBeenCalledTimes(1);
    const calls = mockIntersectionObserver.mock.calls;
    const [onChange] = calls[calls.length - 1];
    onChange([{ isIntersecting: false, intersectionRatio: 0.5 }]);

    await waitFor(() => expect(queryByText("error")).not.toBeInTheDocument());
    await waitFor(() => expect(queryByText("loading")).toBeInTheDocument());

    expect(container).toMatchSnapshot();
  })

  it("bad image renders error", async () => {
    setupImageMock(false, true);
    mockNativeLazyLoadSupportedGetter.mockReturnValue(false);

    const { queryByText } = render(<ImgDynamic
      id={`imgId`}
      loadingComp={loading}
      errorComp={error}
      width={imageSize}
      height={imageSize}
      src={testImg}
    />)


    expect(mockIntersectionObserver).toHaveBeenCalledTimes(1);
    const calls = mockIntersectionObserver.mock.calls;
    const [onChange] = calls[calls.length - 1];
    onChange([{ isIntersecting: true, intersectionRatio: 0.5 }]);

    await waitFor(() => expect(queryByText("error")).toBeInTheDocument());

  })
})