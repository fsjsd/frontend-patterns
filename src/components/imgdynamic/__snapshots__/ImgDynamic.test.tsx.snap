// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImgDynamic does not render image off screen 1`] = `
.c0 {
  position: relative;
}

.c0> * {
  top: 0;
  left: 0;
  position: absolute;
}

<div>
  <div
    class="c0"
  >
    <div>
      loading
    </div>
  </div>
</div>
`;

exports[`ImgDynamic renders image correctly 1`] = `
.c0 {
  position: relative;
}

.c0> * {
  top: 0;
  left: 0;
  position: absolute;
}

<div>
  <div
    class="c0"
  >
    <div>
      loading
    </div>
    <img
      height="200"
      id="imgId"
      loading="lazy"
      src="data:image/gif;base64,R0lGODlhEAAQAMQAAORHHOVSKudfOulrSOp3WOyDZu6QdvCchPGolfO0o/XBs/fNwfjZ0frl3/zy7////wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACH5BAkAABAALAAAAAAQABAAAAVVICSOZGlCQAosJ6mu7fiyZeKqNKToQGDsM8hBADgUXoGAiqhSvp5QAnQKGIgUhwFUYLCVDFCrKUE1lBavAViFIDlTImbKC5Gm2hB0SlBCBMQiB0UjIQA7"
      style="visibility: hidden;"
      width="200"
    />
  </div>
</div>
`;

exports[`ImgDynamic renders image correctly on legacy browsers 1`] = `
.c0 {
  position: relative;
}

.c0> * {
  top: 0;
  left: 0;
  position: absolute;
}

.c0 img {
  -webkit-animation: cJoqxJ ease 0.5s;
  animation: cJoqxJ ease 0.5s;
  -webkit-animation-iteration-count: 1;
  animation-iteration-count: 1;
  -webkit-animation-fill-mode: forwards;
  animation-fill-mode: forwards;
}

<div>
  <div
    class="c0"
  >
    <div>
      loading
    </div>
    <img
      height="200"
      id="imgId"
      src="data:image/gif;base64,R0lGODlhEAAQAMQAAORHHOVSKudfOulrSOp3WOyDZu6QdvCchPGolfO0o/XBs/fNwfjZ0frl3/zy7////wAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACH5BAkAABAALAAAAAAQABAAAAVVICSOZGlCQAosJ6mu7fiyZeKqNKToQGDsM8hBADgUXoGAiqhSvp5QAnQKGIgUhwFUYLCVDFCrKUE1lBavAViFIDlTImbKC5Gm2hB0SlBCBMQiB0UjIQA7"
      width="200"
    />
  </div>
</div>
`;
